### PR TITLE
Move types dependencies to devDependencies

### PR DESCRIPTION
**Pull request by chrisnicola**
This can cause projects when brought into other projects. They should be development dependencies only.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,19 +32,19 @@
         "node": ">=6"
     },
     "dependencies": {
-        "@types/glob": "5 - 7",
-        "@types/parse5": "^5",
         "css-selector-parser": "^1.3",
         "glob": "5 - 7",
         "parse5": "^5",
         "pofile": "^1",
```
Comment (lukasgeiter) [inline package.json]: There's a trailing comma here, which is why the Travis build failed

**Comment by lukasgeiter** [comment]
Hi @chrisnicola, thanks for your PR.

The `typescript` dependency cannot be moved into `devDependencies`. The TypeScript compiler is used for parsing any JS-like sources to extract messages.

Initially I actually had the `@types` in `devDependencies`  but it turned out to cause problems with TypeScript users of the package (#17). Because the types of this package reference external types the TypeScript build would fail if those aren't present.

I realize that including dependencies that not everyone needs by default is not ideal, but I want to avoid requiring users to also manually install those types. In the next major version I will actually split the package into separate packages - one for each language/framework which should improve this situation a bit.

---

What is the problem with these dependencies you are running into?

File: package.json (hunk 1 above)
Comment (chrisnicola) [reply package.json]: Ack sorry, fixing.

**Comment by chrisnicola** [comment]
@lukasgeiter those users should install those types themselves if they are really needed, but they shouldn't be. At least not anymore.

Typescript has improved a lot in this respect and rarely should users ever need to install @types packages just to make Typescript happy (I realize this often used to be the case). They are primarily only for convenience now.

Additionally @types are often are out of date and out of sync with the actual package versions as you might imagine.

Bringing in `@types/node` for example will conflict with anyone who has `@types/requirejs` installed as the both define `require`. This was the specific problem I ran into. By making this a dependency of your project you basically force my project to have it as a dependency too.

I was able to fix this locally by simply removing @types/requirejs as it is no longer necessary for our project to have this installed, again largely because Typescript no longer has this issue.

**Comment by chrisnicola** [comment]
For reference this is the type of error one would encounter:

```
ERROR in .../node_modules/@types/node/module.d.ts
ERROR in .../node_modules/@types/node/module.d.ts(2,5):
TS2300: Duplicate identifier 'mod'.

ERROR in .../node_modules/@types/requirejs/index.d.ts
ERROR in .../node_modules/@types/requirejs/index.d.ts(38,11):
TS2300: Duplicate identifier 'mod'.

ERROR in .../node_modules/@types/requirejs/index.d.ts
ERROR in .../node_modules/@types/requirejs/index.d.ts(422,13):
TS2403: Subsequent variable declarations must have the same type.  Variable 'require' must be of type 'NodeRequire', but here has type 'Require'.
```

**Comment by lukasgeiter** [comment]
Thanks for explaining the issue you experienced. I can see that especially types which define globals can be problematic if installed by default.

---

> Typescript has improved a lot in this respect and rarely should users ever need to install @types packages just to make Typescript happy (I realize this often used to be the case). They are primarily only for convenience now.

I strongly disagree with you on this point. I use TypeScript not so I have to _make it happy_ but rather to help me write correct code and give me accurate code completion. For example, one of the dependencies is `@types/glob` (which in turn includes `@types/node`). Without `@types/glob` this code would compile:

```typescript
let extractor = new GettextExtractor();
extractor.createJsParser().parseFilesGlob('src/**/*.ts', { foo: 'bar' });
```
Of course, `foo` is not a valid option for `node-glob` and if `@types/glob` is installed, TypeScript will actually tell me that.

---

> Additionally @types are often are out of date and out of sync with the actual package versions as you might imagine.

That's true in some cases, but in my experience they still provide a great value.

**Comment by lukasgeiter** [comment]
I'm closing this PR for two reasons:

1. To me this counts as a _somewhat_ breaking change. Therefore it would require a major release. However I'm currently already working on the next major release. The new release will come with a split into separate packages and overhaul of the architecture.
1. I'm not convinced that this is the right solution. I am going to try my best and find a solution that works for everyone in the upcoming major release.

Finally, I want to say that I appreciate you bringing this issue to my attention. Thanks.

**Comment by chrisnicola** [comment]
Thanks for looking into this @lukasgeiter. As I stated I was able to work around this so it wasn't more than a minor inconvenience.

I'm not really clear what disadvantage you see in moving to dev dependencies here though.

Perhaps I'm misunderstanding how this library is built but when I use typescript, I compile the libraries to ES5 or ES6 NPM packages for compatibility that include the type definitions beside them.

To my knowledge this is typically how most typescript projects work, even heavily Typescript based projects like Angular.

I'm not sure I can think of any reason not including types in the dependencies would break someone else compiling.

**Comment by lukasgeiter** [comment]
> Perhaps I'm misunderstanding how this library is built but when I use typescript, I compile the libraries to ES5 or ES6 NPM packages for compatibility that include the type definitions beside them.

That's also how it works in this project. However, the type definitions included in the build output **reference 3rd party types**. One example is [`parser.d.ts`](https://unpkg.com/gettext-extractor@3.4.3/dist/parser.d.ts) which imports `glob` (`@types/glob`) to use the `glob.IOptions` interface for the parameter of `parseFilesGlob`. ([also see comment above](https://github.com/lukasgeiter/gettext-extractor/pull/29#issuecomment-505772083))

> I'm not sure I can think of any reason not including types in the dependencies would break someone else compiling.

I don't think it would break anyone's build. TypeScript seems to handle missing 3rd party types in definitions pretty gracefully. That means it will just treat the missing type as `any`.

The problem I see is someone's build **not breaking** even though there are errors which could be caught by type checking. The package includes type definitions to give TypeScript users full type checking out of the box . Removing those dependencies would change that.

**Comment by chrisnicola** [comment]
@lukasgeiter I have to disagree there, as per my example forcing the downstream consumer to install your type definition choices potentially conflicts with their choices, that seems bad to me.

I'm also not sure why your use of types would necessitate providing the type definitions to consumers. I've looked at the other issue you linked and best I can tell that person simply had typescript misconfigured. You change may have fixed their problem but that isn't entirely clear to me that it was the cause.

I'm also not 100% certain you need @types/glob included to enforce the type checking of your own exported function. I would think the generated d.ts files for this project would do that already. I'd have to test it, but I'd be really surprised if that was a necessary step.

**Comment by lukasgeiter** [comment]
> I have to disagree there, as per my example forcing the downstream consumer to install your type definition choices potentially conflicts with their choices, that seems bad to me.

Because you reported this issue, I know now that it can cause problems for consumers. I agree that the current setup isn't ideal. However I'm trying to explain that your solution has its downsides as well.

> I'm also not 100% certain you need @types/glob included to enforce the type checking of your own exported function. I would think the generated d.ts files for this project would do that already. I'd have to test it, but I'd be really surprised if that was a necessary step.

I've tested it and I am 100% certain that `@types/glob` is needed to allow type checking for the glob options argument. To see why, let's look at the relevant parts of [`parser.d.ts`](https://unpkg.com/gettext-extractor@3.4.3/dist/parser.d.ts):

```typescript
import * as glob from 'glob';
// ...
export declare abstract class Parser<TExtractorFunction extends Function, TParseOptions extends IParseOptions> {
    // ...
    parseFilesGlob(pattern: string, globOptions?: glob.IOptions, options?: TParseOptions): this;
    // ...
}
```

`parseFilesGlob` uses the interface `glob.IOptions` imported from `glob`. The interface isn't included in any of the definition files of this package. Therefore without having `@types/node` installed, TypeScript has no way of knowing that type.

---

Again, I'll give my best to find a better solution for the next major release. For the `glob` options, one possibility could be to copy the interface instead of importing it from `@types/glob`. Although I'm not sure that's a good idea either, since I would have to keep it up-to-date.

**Pull request closed without merging**